### DISCUSSION TITLE
Meaning of the acceleration detected by sensor/framelinacc

### DISCUSSION THREAD

**Discussion by zichunxx** [google-deepmind/mujoco#2011] [Asking for Help]
Hi!

When I learned how to add force-related sensors and collect data, I was confused about the meaning of the acceleration measured by `sensor/framelinacc`. According to the document, 
> This element creates a sensor that returns the 3D linear acceleration of the spatial frame of the object, in global coordinates.

I tried to add this sensor to the box object and check the exact value. Here is the whole manipulation process: 

https://github.com/google-deepmind/mujoco/assets/72594017/e6b450d0-410a-4073-932d-9490126368ae

Then, I collect the acceleration data in the z-direction where the horizontal axis represents the timestep.

![obj_acc](https://github.com/google-deepmind/mujoco/assets/72594017/8eaed776-3dbe-48d9-af17-6bd9c9b5f9df)

Before the object was grasped, the measured acceleration was approximately 9.8 and vertically upward. After the object was grabbed, the direction of the acceleration did not change but the value became smaller. 

So, my confusions are 

1) what does this acceleration (which can also be seen as a kind of force) represent?

2) If this sensor measures the acceleration acting on the frame of the object, why is the acceleration not equal to zero (considering gravity) before grasp?

3) Why does the acceleration become smaller after the grasp?

Thanks in advance!

**Comment by Balint-H** [comment]
Hello, I can try to give a preliminary answer for the first 2 questions
1: Have a look at the way the sensor [calls mj_objectAcceleration](https://github.com/google-deepmind/mujoco/blob/fae1f3e56d5a1df93da510462a17432b7b91b0a0/src/engine/engine_sensor.c#L814) to populate its values, [which in turn processes](https://github.com/google-deepmind/mujoco/blob/fae1f3e56d5a1df93da510462a17432b7b91b0a0/src/engine/engine_support.c#L1576) values from `data->cacc`, which were [calculated recursively from it's parent's](https://github.com/google-deepmind/mujoco/blob/fae1f3e56d5a1df93da510462a17432b7b91b0a0/src/engine/engine_core_smooth.c#L1944) `cacc`

2: Because the `cacc` calculation recurses back to the worldbody frame, and the worldbody frame is modelled as accelerating upwards to represent gravity. The effect of this on the sensor is intentional. Yuval explained it in more detail [here](https://github.com/google-deepmind/mujoco/issues/598#issuecomment-1605406993).

3: I'm not entirely sure. When the data was recorded, was the arm fully settled at the end?

**Comment by zichunxx** [comment]
@Balint-H Thanks for your kind help which helps me with the first two questions.

> 3: I'm not entirely sure. When the data was recorded, was the arm fully settled at the end?

The sensor data is recorded after each `mj_step` like this
```
mujoco.mj_step(model, data, n_steps)
obj_frame_vel_sensordata.append(data.sensor("obj_frame_vel_sensor").data.copy())
```

and the whole end-effector trajectory is finished with the IK controller. 

I'm not sure if the decrease in acceleration measured by `framelinacc` after grasp means that there is relative slippage between the object and the gripper pads or if the way to record the acceleration is wrong.
